Storage/persister/sherlock.h

```diff
@@ -120,66 +132,100 @@ class SherlockStreamPersisterImpl {
     } else {
-      CURRENT_THROW(UnderlyingStreamHasExternalDataAuthorityException());
+      TerminateStreamSubscriptionFromLockedSection();
+      std::lock_guard<std::mutex> lock(stream_publishing_mutex_ref_);
+      publisher_used_ = nullptr;
+      publisher_used_ = stream_->template BecomeFollowingStream<current::locks::MutexLockStatus::AlreadyLocked>();
+      const uint64_t save_replay_index = subscriber_instance_->next_replay_index_;
+      subscriber_instance_ = nullptr;
+      SyncReplayStreamFromLockedSectionOrConstructor(save_replay_index);
     }
   }
 
+  // TODO(dkorolev): `BecomeFollowingStorage` maybe?
+
  private:
-  template <current::locks::MutexLockStatus MLS>
-  void SyncReplayStream(uint64_t from_idx = 0u) {
-    for (const auto& stream_record : stream_used_.Persister().Iterate(from_idx)) {
+  // Invariant: both `subscriber_creator_destructor_mutex_` and `stream_publishing_mutex_ref_` are locked,
+  // or the call is taking place from the constructor.
+  void SyncReplayStreamFromLockedSectionOrConstructor(uint64_t from_idx) {
+    for (const auto& stream_record :
+         stream_->Data()->template Iterate<current::locks::MutexLockStatus::AlreadyLocked>(from_idx)) {
       if (Exists<transaction_t>(stream_record.entry)) {
         const transaction_t& transaction = Value<transaction_t>(stream_record.entry);
-        ApplyMutations<MLS>(transaction);
+        ApplyMutationsFromLockedSectionOrConstructor(transaction, stream_record.idx_ts.us);
       }
     }
   }
 
-  template <current::locks::MutexLockStatus MLS = current::locks::MutexLockStatus::NeedToLock>
-  void ApplyMutations(const transaction_t& transaction) {
-    current::locks::SmartMutexLockGuard<MLS> lock(storage_mutex_ref_);
+  void ApplyMutationsFromLockedSectionOrConstructor(const transaction_t& transaction,
+                                                    std::chrono::microseconds timestamp) {
     for (const auto& mutation : transaction.mutations) {
       fields_update_f_(mutation);
     }
+    PushTimestampToFromLockedSection(timestamp);
   }
 
-  void SubscribeToStream() {
+ private:
+  // Invariant: `master_follower_change_mutex_` is locked, or the call is happening from the constructor.
+  void SubscribeToStreamFromLockedSection() {
     CURRENT_ASSERT(!subscriber_scope_);
-    CURRENT_ASSERT(subscriber_);
-    subscriber_scope_ = std::move(stream_used_.template Subscribe<transaction_t>(*subscriber_));
+    CURRENT_ASSERT(subscriber_instance_);
+    subscriber_scope_ = std::move(stream_->template Subscribe<transaction_t>(*subscriber_instance_));
   }
 
-  void TerminateStreamSubscription() { subscriber_scope_ = nullptr; }
+  // Invariant: `master_follower_change_mutex_` is locked.
+  // Important: The publishing mutex of the respective stream must be unlocked!
+  void TerminateStreamSubscriptionFromLockedSection() { subscriber_scope_ = nullptr; }
+
+  void PushTimestampToFromLockedSection(std::chrono::microseconds timestamp) {
```

Comment:
`SetLastAppliedTimestampFromLockedSection` is longer but cold be less ambiguous imho